pkg/v1/v1util/matcher.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1util
+
+import (
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ManifestMatcher function that is given a v1.Descriptor, and returns whether or
+// not it matches a given rule. Can match on anything it wants in the Descriptor.
+type ManifestMatcher func(desc v1.Descriptor) bool
+
+// NameMatcher returns a ManifestMatcher that matches on the name tag in
+// github.com/opencontainers/image-spec/specs-go/v1.AnnotationRefName
+func NameMatcher(name string) ManifestMatcher {
+	return AnnotationMatcher(imagespec.AnnotationRefName, name)
+}
+
+// AnnotationMatcher returns a ManifestMatcher that matches based on the provided annotation
+func AnnotationMatcher(key, value string) ManifestMatcher {
+	return func(desc v1.Descriptor) bool {
+		if desc.Annotations == nil {
+			return false
+		}
+		if aValue, ok := desc.Annotations[key]; ok && aValue == value {
+			return true
+		}
+		return false
+	}
+}
+
+// PlatformMatcher returns a ManifestMatcher that matches on the provided platform.
+// Ignores any descriptors that do not have a platform.
+func PlatformMatcher(platform v1.Platform) ManifestMatcher {
+	return func(desc v1.Descriptor) bool {
+		if desc.Platform == nil {
+			return false
+		}
+		return platformEqual(*desc.Platform, platform)
+	}
+}
+
+// MediaTypeMatcher returns a ManifestMatcher that matches based on one of the provided media types
+func MediaTypeMatcher(mediaTypes []string) ManifestMatcher {
+	mts := map[string]bool{}
+	for _, media := range mediaTypes {
+		mts[media] = true
+	}
+	return func(desc v1.Descriptor) bool {
+		if desc.MediaType == "" {
+			return false
+		}
+		if _, ok := mts[string(desc.MediaType)]; ok {
+			return true
+		}
+		return false
+	}
+}
+
+// platformEqual return if two v1.Platform are functionally equivalent.
+func platformEqual(a, b v1.Platform) bool {
```

Comment:
For a final implementation, we will actually want to implement proper matching here and document the semantics, but this is good enough for arguing about design for now. I'd like to look at containerd for inspiration. I originally tried to just use their package, but the dependencies and [this](https://github.com/containerd/containerd/blob/4dc698d09b3024ac74948d3551da1fdbdf667200/platforms/cpuinfo.go#L33) drove me away.

The platform matching logic is currently embedded in `remote`, but it should probably just be a method on `Platform` with a reasonable name.

When I implemented this a while ago [for python](https://github.com/google/containerregistry/blob/8a11dc8c53003ecf5b72ffaf035ba280109356ac/client/v2_2/docker_image_list_.py#L43-L107), I came up with:

`a.can_run(b)` returns true if platform `a` has the right os/arch/variant/version and a superset of the features in platform `b`.

I don't like that name :)

Reply:
Oh, yes, even better. I would like to be able to do:

```go
func (p Platform) Equal(other Platform) bool {
  // ...
}
```

If you are good with that, I am more than happy to move it from here to `v1.Platform`; let me know.